src/stream/src/executor/aggregation/agg_state.rs

```diff
@@ -63,46 +75,39 @@ impl<S: StateStore> AggState<S> {
         }
     }
 
-    /// Returns whether `prev_states` is filled.
-    pub fn is_dirty(&self) -> bool {
-        self.prev_states.is_some()
-    }
-
-    /// Used for recording the output of current states as previous states, before applying new
-    /// changes to the state. If the state is already marked dirty in this epoch, this function does
-    /// no-op.
-    /// After calling this function, `self.is_dirty()` will return `true`.
-    pub async fn may_mark_as_dirty(
+    /// Get the outputs of all managed states.
+    async fn get_outputs(
         &mut self,
-        state_tables: &[StateTable<S>],
-    ) -> StreamExecutorResult<()> {
-        if self.is_dirty() {
-            return Ok(());
-        }
-
-        let mut outputs = vec![];
-        for (state, state_table) in self.managed_states.iter_mut().zip_eq(state_tables.iter()) {
-            outputs.push(state.get_output(state_table).await?);
-        }
-        self.prev_states = Some(outputs);
-        Ok(())
+        agg_state_tables: &[Option<AggStateTable<S>>],
+    ) -> StreamExecutorResult<Vec<Datum>> {
+        futures::future::try_join_all(
+            self.managed_states
+                .iter_mut()
+                .zip_eq(agg_state_tables)
+                .map(|(state, agg_state_table)| state.get_output(agg_state_table.as_ref())),
+        )
+        .await
     }
 
-    /// Build changes into `builders` and `new_ops`, according to previous and current states. Note
-    /// that for [`crate::executor::HashAggExecutor`].
+    /// Build changes into `builders` and `new_ops`, according to previous and current agg outputs.
+    /// Note that for [`crate::executor::HashAggExecutor`].
     ///
-    /// Returns how many rows are appended in builders.
+    /// Returns how many rows are appended in builders and the result row including group key
+    /// prefix.
+    ///
+    /// The saved previous outputs will be updated to the latest outputs after building changes.
     pub async fn build_changes(
         &mut self,
         builders: &mut [ArrayBuilderImpl],
         new_ops: &mut Vec<Op>,
-        state_tables: &[StateTable<S>],
-    ) -> StreamExecutorResult<usize> {
-        if !self.is_dirty() {
-            return Ok(0);
-        }
-
-        let row_count = self.row_count();
+        agg_state_tables: &[Option<AggStateTable<S>>],
+    ) -> StreamExecutorResult<(usize, Row)> {
+        let curr_outputs = self.get_outputs(agg_state_tables).await?;
+
+        let row_count = curr_outputs[ROW_COUNT_COLUMN]
+            .as_ref()
+            .map(|x| *x.as_int64())
+            .unwrap_or(0);
```

Comment:
How about panicking here?

Reply:
Yes we can. The row count column must have value here, otherwise `build_changes` shouldn't be called.